programs/address-lookup-table/src/processor.rs

```diff
@@ -0,0 +1,377 @@
+        }
+
+        drop(lookup_table_account_ref);
+        lookup_table_account
+            .try_account_ref_mut()?
+            .set_data(new_table_data);
+
+        let rent: Rent = invoke_context.get_sysvar(&rent::id())?;
+        let required_lamports = rent
+            .minimum_balance(new_table_data_len)
+            .max(1)
+            .saturating_sub(lookup_table_account.lamports()?);
+
+        let table_key = *lookup_table_account.unsigned_key();
+        if required_lamports > 0 {
+            invoke_context.native_invoke(
+                system_instruction::transfer(&payer_key, &table_key, required_lamports),
+                &[payer_key],
+            )?;
+        }
+
+        Ok(())
+    }
+
+    fn close_lookup_table(
+        invoke_context: &mut InvokeContext,
+        first_instruction_account: usize,
+    ) -> Result<(), InstructionError> {
+        let keyed_accounts = invoke_context.get_keyed_accounts()?;
+
+        let lookup_table_account =
+            keyed_account_at_index(keyed_accounts, first_instruction_account)?;
+        if lookup_table_account.owner()? != crate::id() {
+            return Err(InstructionError::InvalidAccountOwner);
+        }
+
+        let authority_account =
+            keyed_account_at_index(keyed_accounts, checked_add(first_instruction_account, 1)?)?;
+        if authority_account.signer_key().is_none() {
+            return Err(InstructionError::MissingRequiredSignature);
+        }
+
+        let recipient_account =
+            keyed_account_at_index(keyed_accounts, checked_add(first_instruction_account, 2)?)?;
+        if recipient_account.unsigned_key() == lookup_table_account.unsigned_key() {
+            ic_msg!(
+                invoke_context,
+                "Lookup table cannot be the recipient of reclaimed lamports"
+            );
+            return Err(InstructionError::InvalidArgument);
+        }
+
+        let lookup_table_account_ref = lookup_table_account.try_account_ref()?;
+        let lookup_table_data = lookup_table_account_ref.data();
+        let lookup_table = AddressLookupTable::deserialize(lookup_table_data)?;
+
+        if lookup_table.meta.authority.is_none() {
+            return Err(InstructionError::Immutable);
+        }
+        if lookup_table.meta.authority != Some(*authority_account.unsigned_key()) {
+            return Err(InstructionError::IncorrectAuthority);
+        }
+
+        // Assert that the slot used in the derivation path of the lookup table address
+        // is no longer recent and can't be reused to initialize an account at the same address.
+        let slot_hashes: SlotHashes = invoke_context.get_sysvar(&slot_hashes::id())?;
+        if let Some(position) = slot_hashes.position(&lookup_table.meta.derivation_slot) {
+            let expiration = MAX_ENTRIES.saturating_sub(position);
+            ic_msg!(
+                invoke_context,
+                "Table cannot be closed until its derivation slot expires in {} blocks",
```

Comment:
Should this be?
```suggestion
                "Table cannot be closed until its derivation slot expires in {} slots",
```

Reply:
It's actually blocks. The recent slot hashes sysvar is only updated for new blocks (https://github.com/solana-labs/solana/blob/622fd7c7ecc8d0756a35e911e30ee5f8af4dceb4/runtime/src/bank.rs#L1932)

Reply:
Gotcha, I had a feeling you were right on this one :smile: